src/codegen/llvm/codegen_cpu.cc

```diff
@@ -562,6 +562,72 @@ llvm::Value* CodeGenCPU::CreateCallPacked(const Call* op) {
   return rvalue;
 }
 
+llvm::Value *CodeGenCPU::CreateCallTracePacked(const Call *op) {
+  using llvm::BasicBlock;
+  CHECK_EQ(op->args.size(), 6U);
+  std::string func_name = op->args[0].as<StringImm>()->value;
+  llvm::Value *handle = GetPackedFuncHandle(func_name);
+  int64_t begin = op->args[3].as<IntImm>()->value;
+  int64_t end = op->args[4].as<IntImm>()->value;
+  int64_t nargs = end - begin;
+  CHECK_GE(nargs, 0);
+  llvm::Value *stack_value = MakeValue(op->args[1]);
+  llvm::Value *stack_tcode = MakeValue(op->args[2]);
+  // Get traced value.
+  llvm::Value *traced_value = MakeValue(op->args[5]);
+  llvm::Value *arg_value = builder_->CreateInBoundsGEP(
+      builder_->CreatePointerCast(stack_value, t_tvm_value_->getPointerTo()),
+      ConstInt32(begin));
+  llvm::Value *arg_tcode =
+      CreateBufferPtr(Int(32), stack_tcode, ConstInt32(begin));
+  llvm::Value *ret_value = builder_->CreateInBoundsGEP(
+      builder_->CreatePointerCast(stack_value, t_tvm_value_->getPointerTo()),
+      ConstInt32(end));
+  llvm::Value *ret_tcode =
+      CreateBufferPtr(Int(32), stack_tcode, ConstInt32(end));
+  BasicBlock *end_block = CheckCallSuccess(builder_->CreateCall(
+      RuntimeTVMFuncCall(),
+      {handle, arg_value, arg_tcode, ConstInt32(nargs), ret_value, ret_tcode}));
+  // The update_block handles case when we need to update the return value.
+  BasicBlock *update_block =
+      BasicBlock::Create(*ctx_, "update_block", function_);
+  // The continue_block handles case when we need to return original
+  // traced value.
+  BasicBlock *continue_block =
+      BasicBlock::Create(*ctx_, "continue_block", function_);
+  // Load the return value.
+  Type r_type = op->type;
+  // Have to convert to 64 bit type, because TVMValue is 8 byte long.
+  Type r_api_type = ir::APIType(r_type);
+  llvm::Value* rvalue =
+      builder_->CreateAlignedLoad(
+          builder_->CreatePointerCast(
+              ret_value, LLVMType(r_api_type)->getPointerTo()), 8);
+  // Convert back to op type.
+  rvalue = CreateCast(r_api_type, r_type, rvalue);
+  // Check the return value type and create cmp instruction.
+  llvm::Value *cmp;
+  llvm::Type *rvalue_type = rvalue->getType();
+  // The cmp instruction depends on processing type.
+  if (rvalue_type->isFloatTy() || rvalue_type->isDoubleTy() ||
```

Comment:
thanks for mention tcode, I was spending couple of hours to understand how to check return value, and didn't find a better solution to return 0 from call back when we want to process origin value, otherwise return user-defined value, but the true way to do it is by checking tcode.